src/Aacotroneo/Saml2/Http/Controllers/Saml2Controller.php

```diff
@@ -100,4 +103,146 @@ public function login()
+    private function getHL7MessageFromRequest( $user )
+    {
+        $attributes = $user->getAttributes();
+
+        if( !array_key_exists('RequestMessage', $attributes) || sizeof($attributes['RequestMessage']) === 0 )
+        {
+            return null;
+        }
+
+        return $attributes['RequestMessage'][0];
+    }
+
+    /**
+     * Get the user model using the email on request
+     *
+     * @param   Saml2User   $user
+     * @return  User
+     */
+    private function getUserFromRequest( $samlUser )
+    {
+        $email = $samlUser->getAttributes()['Email'][0];
+
+        // Find a node with the attribute Name set as Email, after find the text node that contains the email
+        $user = User::whereUsername( $email )->first();
+        if ( is_null($user) )
+        {
+            throw new Saml2UserNotPresentException( "Saml request with username {$username} does not have an user on AristaMD" );
+        }
+        return $user;
+    }
+
+    /**
+     * Get the query params for the request object
+     *
+     * @param   Request   $request
+     * @return  String
+     */
+    public function getUrlParamString( $request )
+    {
+        if( empty($request) || empty($request->id) )
+        {
+            return "";
+        }
+
+        $recordType = null;
+
+        switch ( get_class($request) )
+        {
+            case 'App\Models\EConsult':
+                $recordType = 'econsult';
+                break;
+            case 'App\Models\Referral':
+                $recordType = 'referral';
+                break;
+            default:
+                $recordType = null;
+                break;
+        }
+
+        return "&record_type=$recordType&record_id=$request->id";
+    }
+
+    /**
+     * Creates a referral or an eConsult from a HL7 message
+     *
+     * @param   String  $message    HL7 message as string
+     * @param   User    $user       Authenticated user
+     * @return  Integer             Request id
+     */
+    private function createRequest( $message, $user )
+    {
+        if ( empty($message) )
+        {
+            /*$this->responseCode( 500 );
+            $this->message = "Internal Error. Please provide the HL7 v.2.3 REF message.";
+            return $this->buildResponse( );*/
+        }
+        // Getting the patient data from the HL7 message
+        $patient = HL7::transformPatient( $user->organization->id, $message );
+        $requestObject = HL7::transformRequest( $user->id, $user->organization->id, $message );
+
+        $patient->save();
+
+        $requestObject->patient_id = $patient->id;
+        $requestObject->patient_version = $patient->version;
+
+        $requestObject->save();
+
+        return $requestObject;
+    }
+
+    /**
+     * Checks if there is an HL7 message, process the message and returns the query parameter string
+     *
+     * @param   Saml2User   $user
+     * @return  String
+     */
+    private function processRequestMessage( $user, $redirectUrl )
```

Comment:
I think that we could move the HL7 processing to a helper to avoid the dependency in case this controller has to change to include another functionality